### PR TITLE
Auto flashing without asking permissions if possible

### PR DESCRIPTION
**Pull request by McGiverGim**
This enables the possibility of flashing without asking for permissions. There are several cases:
1. If the user has gived permissions before and the board is in DFU mode (the port picker has a DFU port selected), it simply starts flashing.
2. If the user has gived permissions before for DFU and the board is in serial mode (the port picker has a SERIAL port selected), the webstm32 reboots in DFU and lets the firmware_flasher detect the new device and flash it.
3. If the user has NOT gived permissions before for DFU and the board is in serial mode (the port picker has a SERIAL port selected), the webstm32 reboots in DFU and after a timeout (to see if the firmware_flasher starts the flashing), shows the permissions window and lets the firmware_flasher detect the new device and flash it.
4. If there is not any DFU or SERIAL device selected in the port_handler, maybe the board is in DFU mode but without permissions, so it asks for permissions and starts the flashing.
5. If the auto-flash is enabled, and a DFU board is detected, it simply starts the flashing.

I've observed one problem, but it occurs too before this PR, that if the user don't select a DFU device when asking for permissions, the flashing buttons remain disabled. As I say it happens before too, so I have not fixed it here.

I've tested a lot of flashing, but maybe there are some edge cases that need some little modification. I've observed other problems before this PR too, like the auto-detect appearing as disabled sometimes.

### CODE REVIEW

**Comment by nerdCopter** [comment]
`f3d12d22`
- tested successful: 1 try. `detect` FC, unlplug. `load firmware`, hold-boot-button & plug, immediate flash.

**Review by nerdCopter — APPROVED** [review]

**Comment by haslinghuis** [comment]
@nerdCopter Still need to click Flash Firmware button? Otherwise it does not start

**Comment by nerdCopter** [comment]
> @nerdCopter Still need to click Flash Firmware button? Otherwise it does not start

i actually flashed twice, once using `flash` button, then once with only `load firmware` ; i will retest, maybe `flash` got it into a kept-state :shrug:

**Comment by nerdCopter** [comment]
unplug, `load firmware`, `flash on connect`, plug as DFU --> immediate.

**Comment by haslinghuis** [comment]
Yes but ONLY when flashing a second time - i.e. the port has been granted permission (this session) it works without new permission request.

Perhaps rather we remove flash on connect as each board needs it's own permission as the permission dialog acts as a timeout waiting for DFU to show up.

**Comment by nerdCopter** [comment]
yes, and perhaps yes.

**Comment by McGiverGim** [comment]
> Perhaps rather we remove flash on connect as each board needs it's own permission as the permission dialog acts as a timeout waiting for DFU to show up.

I don't understand exactly... The flash on connect does not have any timeout. The timeout for the permission request it's used only when rebooting from serial to dfu. 

Maybe it's not too useful now, but it does not hurt, except if there is some bug, of course.

File: src/js/protocols/webstm32.js [hunk 1]
```diff
@@ -100,10 +100,31 @@ class STM32Protocol {
...
+                            GUI.connect_lock = false;
+                        }
+                    })
+                    .catch (e => {
+                        console.error('DFU request permission failed', e);
+                        this.rebootMode = 0;
+                        GUI.connect_lock = false;
+                    });
+                }
+            }, 3000);
```
Comment (haslinghuis) [inline src/js/protocols/webstm32.js:126]: Chris needed 5 seconds here (in NWjs version)

**Review by haslinghuis — APPROVED** [review]

File: src/js/protocols/webstm32.js (hunk 1 above)
Comment (McGiverGim) [reply src/js/protocols/webstm32.js:126]: Interesting. Here is almost immediate. I let enough time but not too much for the user that is waiting...

Comment (haslinghuis) [reply src/js/protocols/webstm32.js:126]: Guess it was without flash on connect. Not sure how much it is needed anymore as I have never needed or used it before. Afaik it was used on F3 which we no longer support. We don't track the use of it so we just don't know :)

**Pull request merged**